### PR TITLE
specify common arch for mkldnn build, so it can be reused on all arch

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##

specify common arch for mkldnn build, so it can be reused on all arch

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: prepare_mkldnn.sh [hunk 1]
```diff
@@ -93,7 +93,7 @@ if [ ! -f $MKLDNN_LIBFILE ]; then
     echo "Building MKLDNN ..." >&2
     cd $MXNET_ROOTDIR
 	g++ --version >&2
-    cmake $MKLDNN_ROOTDIR -DCMAKE_INSTALL_PREFIX=$MKLDNN_INSTALLDIR -B$MKLDNN_BUILDDIR -DARCH_OPT_FLAGS="-mtune=generic" >&2
+    cmake $MKLDNN_ROOTDIR -DCMAKE_INSTALL_PREFIX=$MKLDNN_INSTALLDIR -B$MKLDNN_BUILDDIR -DARCH_OPT_FLAGS="-mtune=generic -march=haswell" >&2
```
[the remark this answers is not in the document]
Comment (ashokei) [reply prepare_mkldnn.sh]: This should not impact mkldnn primitive execution, which are generated runtime based on underly arch.

[the remark this answers is not in the document]
Comment (szha) [reply prepare_mkldnn.sh]: we have received reports before from users who don't even have AVX2. Please don't assume haswell.

[the remark this answers is not in the document]
Comment (ashokei) [reply prepare_mkldnn.sh]: `ivybridge` sounds like a good compromise. 

https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html 

@szha does that look ok, thanks.

**Comment by szha** [comment]
Why is arch needed? cc'd @zheng-da

**Comment by ashokei** [comment]
If we build on same machine, and use on same machine then there is no problem. I notice CI , or other use-cases where we may be building on one machine and copying mkldnn lib along with mxnet to other machine with differnt arch. This PR will make the build be compatible by using a common baseline.

There will not be any performance impact, as the mkldnn kernels/primitives are generated runtime based on underlying arch. This PR is only to make the build itself compatible across systems.

**Comment by TaoLv** [comment]
I‘m afraid this will still imapct the performance of non-jit code if arch is hard coded here. Can we use a flag like `NATIVE` to control compilation during make? If the flag is True, we should compile mxnet/mkldnn according to the native machine. Otherwise, only a minimum instruction set should be supported. Anyway, the flag would be default to True.

@szha Do we need to define a subset of instruction or arch which mxnet should support? I notice that there is also a discussion on dev@ about removing ancient gpu arch support from mxnet.

**Comment by szha** [comment]
@TaoLv that means we need to at least compromise the performance of the pre-built binary, and I'd rather we don't go down that route. can we instead make a similar option to DYNAMIC_ARCH in openblas?

**Comment by zheng-da** [comment]
@ashokei What you described is basically what `-mtune=generic` can offer. You explained it to me before, but I forgot now. What benefit does `-march=ivybridge` offer?
And what if the architecture isn't ivybridge?

**Comment by ashokei** [comment]
@TaoLv @szha I measured the performance with this PR change on inference models using benchmark_score script. I did not see any performance difference. 

Intel mkl-dnn team recommended this approach when i asked them about "similar option to DYNAMIC_ARCH in openblas" for mkldnn.  

mklml library (which is shipped as part of mkldnn) uses this approach.

@zheng-da This flag will ensure a common baseline arch is used, instead of `native` or default arch of the system used for building mkldnn. This will just ensure the auxiliary/support code does not cause illegal instruction/crash etc. The actual mkldnn kernels/primitives are generated at runtime based on underlying arch of the system being used for running.

**Comment by ashokei** [comment]
we can explore other options.  One solution is to do this if some environment variable is exported, maybe. @szha is that ok? we could use it only for CI , or other use-cases where users may build on one system and copy to different arch system.

**Comment by szha** [comment]
the bottomline is, the current solution happens at build-time. build-time environment is different from run-time environment. if the integration code is not performance-critical, we should make sure the specified option is supported on any arch that mxnet supports. if it is, then a patch for run-time switching is necessary.

**Comment by TaoLv** [comment]
@szha Just to add, I greped in mxnet project and found there is a flag `USE_SSE`. Not sure about the usage of it, but I think there already has some constraints of instruction set for mxnet.  Another thing is, mxnet provides different pip packages with different version of cuda to support different gpu archs. Shall we go this way?

**Comment by szha** [comment]
Hi all. Since the script on which this patch was proposed no longer exist, I'm closing this PR. If this fix is still relevant or you'd like further discussion on the approach, feel free to open another PR/issue. Thanks.

**Pull request closed without merging**